src/coniferest/session/oracle.py

```diff
@@ -0,0 +1,93 @@
+import numpy as np
+from typing import List, Optional
+
+from coniferest.coniferest import Coniferest
+from coniferest.label import Label
+from coniferest.session import Session
+from coniferest.session.callback import TerminateAfter, TerminateAfterNAnomalies
+
+
+class Tracer:
+    def __init__(self):
+        self.trace = []
+
+    def __call__(self, _metadata, _data, session):
+        ordering = np.argsort(session.scores)
+        index = np.where(ordering == session.current)[0].item()
+        self.trace.append(ordering[:index + len(session.known_anomalies) + 1])
+
+
+class OracleSession(Session):
+    """Automated session to run experiments with labeled data.
+
+    Parameters
+    ----------
+    data : np.ndarray, 2-D
+        Array with feature values of objects.
+    labels : np.ndarray, 1-D
+        Array with true labels, of Label or int type
+    model : Coniferest
+        Anomaly detection model to use
+    max_asked : int
+        Maximum number of asked decisions
+    max_anomalies : int
+        Maximum number of anomalies to search for
+    trace : bool or None, optional
+        Whether to save the trace of the experiment. Default is True.
+
+    Attributes
+    ----------
+    trace : list of np.ndarray
+        Trace of the experiment, if trace=True
+
+    Also see methods and attributes from the base `Session` class
+    """
+
+    def __init__(self, data: np.ndarray, labels: np.ndarray, *, model: Coniferest, max_asked: int, max_anomalies: int, trace: bool = True):
+        on_decision_callbacks = [
+            TerminateAfter(max_asked),
+            TerminateAfterNAnomalies(max_anomalies),
+        ]
+        self._tracer = None
+        if trace:
+            self._tracer = Tracer()
+            on_decision_callbacks.insert(0, self._tracer)
+
+        super().__init__(
+            data=data,
+            metadata=labels,
+            model=model,
+            # Session.metadata is labels, so we just use this candidate metadata is the true label
+            decision_callback=lambda label, _features, _self: label,
+            on_decision_callbacks=on_decision_callbacks,
+        )
+
+    @property
+    def trace(self) -> List[np.ndarray[int]]:
+        if self._tracer is None:
+            raise AttributeError("Construct with tracer=True to use .trace")
+        return self._tracer.trace
```

Comment:
Since trace is a kind of public attribute, can we document what trace is?

Reply:
I removed Tracer